### PR TITLE
feat: add GLOBAL_VARIABLE table and support ShowGlobalVariables

### PR DESCRIPTION
**Pull request by keyu813**
* **Please check if the PR fulfills these requirements**

- The commit message follows our guidelines
- Tests for the changes have been added (for bug fixes / features)
- Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature:
add system table global_variable(in information_schema db as same as mysql)
support showglobalvariables in cli
* **What is the current behavior?** (You can also link to an open issue here)
image
* **What is the new behavior (if this is a feature change)?**
image
image

### CODE REVIEW

File: src/nameserver/system_table_test.cc [hunk 1]
```diff
@@ -72,6 +72,20 @@ TEST_F(SystemTableTest, SystemTable) {
     ASSERT_TRUE(ns_client.ShowTable(PRE_AGG_META_NAME, INTERNAL_DB, false, tables, msg));
     ASSERT_EQ(1, tables.size());
     ASSERT_STREQ("PRE_AGG_META_INFO", tables[0].name().c_str());
+
```
Comment (aceforeverd) [inline src/nameserver/system_table_test.cc:78]: `tables.clear()` first

File: src/cmd/sql_cmd.h [hunk 2]
```diff
@@ -754,7 +754,11 @@ void HandleCmd(const hybridse::node::CmdPlanNode* cmd_node) {
             break;
         }
         case hybridse::node::kCmdShowGlobalVariables: {
-            std::cout << "ERROR: global variable is unsupported now" << std::endl;
+            std::string db_ = "INFORMATION_SCHEMA";
+            std::string sql = "select * from GLOBAL_VARIABLE";
+            ::hybridse::sdk::Status status;
+            auto rs = sr->ExecuteSQL(db_, sql, &status);
```
Comment (aceforeverd) [inline src/cmd/sql_cmd.h]: check status first

File: src/cmd/sql_cmd.h [hunk 3]
```diff
@@ -754,7 +754,11 @@ void HandleCmd(const hybridse::node::CmdPlanNode* cmd_node) {
             break;
         }
         case hybridse::node::kCmdShowGlobalVariables: {
-            std::cout << "ERROR: global variable is unsupported now" << std::endl;
+            std::string db_ = "INFORMATION_SCHEMA";
```
Comment (aceforeverd) [inline src/cmd/sql_cmd.h]: need test cover

File: src/nameserver/name_server_impl.cc [hunk 4]
```diff
@@ -5456,6 +5456,11 @@ void NameServerImpl::OnLocked() {
             LOG(FATAL) << "create internal database failed";
             exit(1);
         }
+        status = CreateDatabase(INFORMATION_SCHEMA_DB);
+        if (!status.OK() && status.code != ::openmldb::base::ReturnCode::kDatabaseAlreadyExists) {
+            LOG(FATAL) << "create information schema database failed";
+            exit(1);
+        }
```
Comment (aceforeverd) [inline src/nameserver/name_server_impl.cc]: pls resolve the cpplint problem here and below

File: src/nameserver/system_table.h [hunk 5]
```diff
@@ -33,22 +33,27 @@ constexpr const char* INTERNAL_DB = "__INTERNAL_DB";
 constexpr const char* JOB_INFO_NAME = "JOB_INFO";
 constexpr const char* PRE_AGG_META_NAME = "PRE_AGG_META_INFO";
 
+
+constexpr const char* INFORMATION_SCHEMA_DB = "INFORMATION_SCHEMA";
+constexpr const char* GLOBAL_VARIABLE_NAME = "GLOBAL_VARIABLE";
```
Comment (aceforeverd) [inline src/nameserver/system_table.h]: `GLOBAL_VARIABLES`

File: src/nameserver/name_server_impl.cc [hunk 6]
```diff
@@ -5445,7 +5445,7 @@ void NameServerImpl::OnLocked() {
     if (!Recover()) {
         PDLOG(WARNING, "recover failed");
     }
-    if (IsClusterMode() && (db_table_info_.empty() || db_table_info_[INTERNAL_DB].empty())) {
+    if (IsClusterMode() && (db_table_info_.empty() || db_table_info_[INTERNAL_DB].empty() || db_table_info_[INFORMATION_SCHEMA_DB].empty())) {
```
Comment (aceforeverd) [inline src/nameserver/name_server_impl.cc]: what if INTERNAL_DB exist already and INFORMATION_SCHEMA is not, will this fatal ?

**Review by aceforeverd — CHANGES_REQUESTED** [review]

File: src/nameserver/system_table.h [hunk 7]
```diff
@@ -33,22 +33,27 @@ constexpr const char* INTERNAL_DB = "__INTERNAL_DB";
 constexpr const char* JOB_INFO_NAME = "JOB_INFO";
 constexpr const char* PRE_AGG_META_NAME = "PRE_AGG_META_INFO";
 
+
+constexpr const char* INFORMATION_SCHEMA_DB = "INFORMATION_SCHEMA";
```
Comment (dl239) [inline src/nameserver/system_table.h:37]: why add new database

Comment (keyu813) [reply src/nameserver/system_table.h:37]: > why add new database

GLOBAL_VARIABLES in mysql is stored in db INFORMATION_SCHEMA

Comment (aceforeverd) [reply src/nameserver/system_table.h:37]: we're creating table `information_schema.global_variables` the same as mysql

**Comment by aceforeverd** [comment]
here is next problem to consider, maybe in next pull request !

we now have set session functionality for variable `execute_mode` and `enable_trace`, so they should exists default on the `global_variabes` table, as the startup default value.

**Comment by dl239** [comment]
> we now have set session functionality for variable `execute_mode` and `enable_trace`, so they should exists default on the `global_variabes` table, as the startup default value.

session variables need not store in gloable_variables table. When SDK startup , they will be inited.

**Review by dl239 — APPROVED** [review]

File: src/cmd/sql_cmd_test.cc [hunk 8]
```diff
@@ -271,6 +271,7 @@ TEST_P(DBSDKTest, deploy_options) {
     auto cli = GetParam();
     cs = cli->cs;
     sr = cli->sr;
+    HandleSQL("show global variables;");
```
Comment (aceforeverd) [inline src/cmd/sql_cmd_test.cc]: it should appear in a newly test case

File: src/cmd/sql_cmd.h (hunk 2 above)
Comment (aceforeverd) [reply src/cmd/sql_cmd.h]: check `status` variable .

File: src/cmd/sql_cmd.h [hunk 9]
```diff
@@ -754,7 +754,18 @@ void HandleCmd(const hybridse::node::CmdPlanNode* cmd_node) {
...
+            std::string sql = "select * from GLOBAL_VARIABLES";
+            ::hybridse::sdk::Status status;
+            auto rs = sr->ExecuteSQL(db_, sql, &status);
+            if (!rs) {
```
Comment (aceforeverd) [inline src/cmd/sql_cmd.h]: the check might replaced by `status.code != 0`

File: src/nameserver/system_table_test.cc [hunk 10]
```diff
@@ -59,7 +59,7 @@ TEST_F(SystemTableTest, SystemTable) {
     ::openmldb::client::NsClient ns_client("127.0.0.1:6530", "");
     ns_client.Init();
     std::vector<::openmldb::nameserver::TableInfo> tables;
-    std::string msg;
+    std::string msg;
```
Comment (aceforeverd) [inline src/nameserver/system_table_test.cc]: trailing whitespace

File: src/nameserver/name_server_impl.cc [hunk 11]
```diff
@@ -5445,25 +5445,43 @@ void NameServerImpl::OnLocked() {
...
             LOG(FATAL) << "tablet num " << tablets_.size() << " is less then system table replica num "
                        << FLAGS_system_table_replica_num;
             exit(1);
         }
-        auto status = CreateDatabase(INTERNAL_DB);
-        if (!status.OK() && status.code != ::openmldb::base::ReturnCode::kDatabaseAlreadyExists) {
-            LOG(FATAL) << "create internal database failed";
-            exit(1);
+        if (db_table_info_.empty()) {
+            auto status = CreateDatabase(INTERNAL_DB);
```
Comment (aceforeverd) [inline src/nameserver/name_server_impl.cc:5455]: should check `INTERNAL_DB` & `INFORMATION_SCHEMA` database separately.

**Review by aceforeverd — APPROVED** [review]

**Comment by aceforeverd** [comment]
> > we now have set session functionality for variable `execute_mode` and `enable_trace`, so they should exists default on the `global_variabes` table, as the startup default value.
> 
> session variables need not store in gloable_variables table. When SDK startup , they will be inited.

usually they should, as the default startup option. session variables cover that global variable in that session.

File: src/nameserver/name_server_impl.cc (hunk 11 above)
Comment (aceforeverd) [reply src/nameserver/name_server_impl.cc:5455]: request change :)

**Review by aceforeverd — CHANGES_REQUESTED** [review]

**Review by aceforeverd — APPROVED** [review]

**Comment by keyu813** [comment]
[global variable support](https://trello.com/c/Nqn4zAcQ/29-global-variable-support)

**Comment by keyu813** [comment]
[global variable support](https://trello.com/c/Nqn4zAcQ/29-global-variable-support)

**Comment by keyu813** [comment]
[global variable support](https://trello.com/c/Nqn4zAcQ/29-global-variable-support)

**Comment by keyu813** [comment]
[global variable support](https://trello.com/c/Nqn4zAcQ/29-global-variable-support)

**Pull request merged**